tardis/model/matter/composition.py

```diff
@@ -0,0 +1,180 @@
+        new_nuclide_index.names = ["atomic_number", "mass_number"]
+        return new_nuclide_index
+
+    @property
+    def isotopic_mass_fraction(self):
+        filtered_nuclide_mass_fraction = self.nuclide_mass_fraction[
+            self.nuclide_mass_fraction.index.get_level_values(1) != -1
+        ]
+        return IsotopicMassFraction(filtered_nuclide_mass_fraction)
+
+    @property
+    def elemental_mass_fraction(self):
+        return self.nuclide_mass_fraction.groupby(level=0).sum()
+
+    @property
+    def element_masses(self):
+        """Atomic mass of elements in each shell"""
+        element_masses = self.nuclide_masses[
+            self.nuclide_masses.index.get_level_values(1) == -1
+        ]
+        element_masses.index = element_masses.index.droplevel(1)
+        return element_masses
+
+    @property
+    def effective_element_masses(self):
+        # This is functionality that we will likely want to remove
+        effective_element_masses = self.nuclide_mass_fraction[
+            self.nuclide_mass_fraction.index.get_level_values(1) == -1
+        ].copy()
+        effective_element_masses.index = (
+            effective_element_masses.index.droplevel(1)
+        )
+        for col in effective_element_masses.columns:
+            effective_element_masses[col] = self.element_masses.loc[
+                effective_element_masses.index
+            ]
+
+        current_isotope_masses = ISOTOPE_MASSES.loc[
+            self.isotopic_mass_fraction.index
+        ]
+        contributing_isotope_masses = (
+            self.isotopic_mass_fraction.multiply(current_isotope_masses, axis=0)
+            .groupby(level=0)
+            .sum()
+        )
+        effective_isotopes_masses = (
+            contributing_isotope_masses
+            / self.isotopic_mass_fraction.groupby(level=0).sum()
+        ) * u.u.to(u.g)
+
+        effective_element_masses = pd.concat(
+            [effective_element_masses, effective_isotopes_masses]
+        )
+        return effective_element_masses
+
+    @property
+    def elemental_number_density(self):
+        """Elemental Number Density computed using the formula: (elemental_mass_fraction * density) / atomic mass"""
+        return (
+            self.elemental_mass_fraction
+            * self.density.to(u.g / u.cm**3).value
+        ).divide(
+            self.effective_element_masses.reindex(
+                self.elemental_mass_fraction.index
+            ),
+            axis=0,
+        )
+
+    @property
+    def isotopic_number_density(self):
+        """Isotopic Number Density computed using the formula: (isotopic_mass_fraction * density) / atomic mass"""
+        return (
+            self.isotopic_mass_fraction * self.density.to(u.g / u.cm**3).value
+        ).divide(
+            ISOTOPE_MASSES.loc[self.isotopic_mass_fraction.index] * u.u.to(u.g),
+            axis=0,
+        )
+
+    def calculate_mass_fraction_at_time(self, time_explosion):
```

Comment:
I would prefer this to just be a `time` variable instead of `time_explosion`. Also lacking docstring.

Reply:
I'm not so happy with the `time` as I think there might be multiple times at play. but added docstring

Reply:
Multiple times for this one function?